src/bases/finitebasis.jl

```diff
@@ -0,0 +1,60 @@
+import Base: length,
+    size
+
+###############
+# FiniteBasis #
+###############
+    # A FiniteBasis repesents the tensor product
+    # of an arbitrary number of fock bases, each of which is truncated 
+    # at a known, finite number of states. For example, imagine the 
+    # following fock bases and their tensor product:
+    #
+    # F_4 => {|0>, |1>, |2>, |3>}
+    # F_3 => {|0>, |1>, |2>}
+    # F_2 => {|0>, |1>} 
+    #
+    # F_4 x F_3 x F_2 => {|0,0,0>, |0,0,1>, |0,1,0>, |0,1,1>...|3,2,1>}
+    # 
+    # Note that the subscript number refers to basis *length*, not max
+    # excitation level.
+    #
+    # This tensor product basis, F_4 x F_3 x F_2, could be represented 
+    # by calling FiniteBasis(4, 3, 2). 
+
+    immutable FiniteBasis{S} <: AbstractBasis{S}
```

Comment:
I haven't, but it seems that it pulls some tricks to allow you to encode the length of the tuple in the type. This theoretically makes sense, since a tuple type in and of itself can be a data type, but in practical use it's hard to dispatch on, since the compiler often treats tuple types as values when you're acting on them. For example:
```
julia> type Foo{N}
           tup::Tuple
       end

julia> Foo{T<:Tuple}(tup::T) = Foo{length(T)}(tup)

julia> [Foo((1,2,3)) for i=1:2]
2-element Array{Foo{N},1}:
 Foo{3}((1,2,3))
 Foo{3}((1,2,3))
```
While the outer constructor above is theoretically type stable, Julia's compiler seems to have trouble inferencing types constructed in this manner - as you can see, it couldn't guess the proper element type for that array. I believe this is because `T<:Tuple` is considered to be a "value"  rather than a "type" when consumed by the `length` function. With `NTuple`s, however, one can reliably encode this information:

```
julia> type Bar{N}
           tup::NTuple{N}
       end

julia> [Bar((1,2,3)) for i=1:2]
2-element Array{Bar{3},1}:
 Bar{3}((1,2,3))
 Bar{3}((1,2,3))
```